fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,544 @@
+    ) -> Self
+    where
+        OverseerError: From<DB::Error>,
+    {
+        let thread_is_enabled = is_enabled;
+        let stop_requested = Arc::new(AtomicBool::new(false));
+        let thread_stop_requested = stop_requested.clone();
+        let grpcio_env = Arc::new(grpcio::EnvBuilder::new().build());
+        let ingest_clients: Vec<FogIngestGrpcClient> = ingest_cluster_uris
+            .iter()
+            .map(|fog_ingest_uri| {
+                FogIngestGrpcClient::new(
+                    fog_ingest_uri.clone(),
+                    Self::GRPC_RETRY_SECONDS,
+                    grpcio_env.clone(),
+                    logger.clone(),
+                )
+            })
+            .collect();
+        let join_handle = Some(
+            ThreadBuilder::new()
+                .name("OverseerWorker".to_string())
+                .spawn(move || {
+                    OverseerWorkerThread::start(
+                        ingest_clients,
+                        recovery_db,
+                        thread_is_enabled,
+                        thread_stop_requested,
+                        logger,
+                    )
+                })
+                .expect("Could not spawn OverseerWorkerThread"),
+        );
+
+        Self {
+            join_handle,
+            stop_requested,
+        }
+    }
+
+    /// Stop and join the db poll thread
+    pub fn stop(&mut self) -> Result<(), ()> {
+        if let Some(join_handle) = self.join_handle.take() {
+            self.stop_requested.store(true, Ordering::SeqCst);
+            join_handle.join().map_err(|_| ())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for OverseerWorker {
+    fn drop(&mut self) {
+        let _ = self.stop();
+    }
+}
+
+/// The thread that performs the Fog Overseer logic.
+struct OverseerWorkerThread<DB: RecoveryDb> {
+    /// The list of FogIngestClients that Overseer uses to communicate with
+    /// each node in the Fog Ingest cluster that it's monitoring.
+    ingest_clients: Vec<FogIngestGrpcClient>,
+
+    /// The database that contains, among other things, info on the Fog Ingest
+    /// cluster's ingress keys.
+    recovery_db: DB,
+
+    /// If this is true, the worker will not perform it's monitoring logic.
+    is_enabled: Arc<AtomicBool>,
+
+    /// If this is true, the thread will stop.
+    stop_requested: Arc<AtomicBool>,
+
+    logger: Logger,
+}
+
+/// This associates an IngestSummary with an IngestClient. This makes it easy
+/// to query a given node based on its IngestSummary.
+struct IngestSummaryNodeMapping {
+    node_index: usize,
+
+    ingest_summary: IngestSummary,
+}
+
+impl<DB: RecoveryDb> OverseerWorkerThread<DB>
+where
+    OverseerError: From<DB::Error>,
+{
+    /// Poll the Fog Ingest cluster every 5 seconds.
+    const POLLING_FREQUENCY: Duration = Duration::from_secs(5);
+
+    /// Try a request to Fog Ingest node this many times if you encounter an
+    /// error.
+    const NUMBER_OF_TRIES: u64 = 3;
```

Comment:
Is there any reason for this to be a `u64` instead of a `usize`? It would let you remove all the `try_into().unwrap()` calls later...

Reply:
Yes, the reason is that the `current_try` parameter is `u64` (kinda weird to me that the retry library defines the types this way). 

I was talking with eran about this very concern and weakly prefer this to writing `as` when I do the subtraction to see how many calls remain (see line 523). that said, will change it if you'd like.

Reply:
I'm ambivalent, but there's definitely a strong preference against `unwrap()` dating back to the early days of our codebase ;-)

I suppose the main thing is: `usize as u64` is infallible on all platforms today, whereas (obviously) `try_into().unwrap()` is not.

Reply:
Thanks for the context. Didn't realize that last point. Implemented it as you suggested. Done.